### PR TITLE
Add R HTTR Code Generator

### PR DESCRIPTION
**Pull request by erfemega**
Adding r-httr codegenerator

### CODE REVIEW

File: codegens/r-httr/README.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+
+> Converts Postman-SDK Request into code snippet for .
+
+#### Prerequisites
+To run Code-Gen, ensure that you have NodeJS >= v8. A copy of the NodeJS installable can be downloaded from https://nodejs.org/en/download/package-manager.
+
+## Using the Module
+The module will expose an object which will have property `convert` which is the function for converting the Postman-SDK request to swift code snippet.
```
Comment (LuisTejedaS) [inline codegens/r-httr/README.md:8]: This should say to r code snippet instead of swift code

File: codegens/r-httr/lib/rHttr.js [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+/**
+ * Transforms an array of headers into the desired form of the language
+ *
+ * @param  {Array} mapToSnippetArray - array of key values
+ * @param  {String} indentation - used for indenting snippet's structure
+ * @param  {boolean} sanitize - whether to sanitize the key and values
+ * @returns {String} - array in the form of [ key => value ]
+ */
+function getSnippetArray (mapToSnippetArray, indentation, sanitize) {
+  // mapToSnippetArray = groupHeadersSameKey(mapToSnippetArray);
```
Comment (LuisTejedaS) [inline codegens/r-httr/lib/rHttr.js:92]: You may forget to add the group headers or this commented line is no longer needed

File: codegens/r-httr/lib/rHttr.js [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+function getSnippetRequest ({url, method, hasParams, hasHeaders, mode, requestTimeout}) {
+  const methodUC = method.toUpperCase();
+  let snippetRequest = '';
+  if (methodUC === 'GET' || methodUC === 'DELETE') {
+    snippetRequest = getSnippetGetOrDeleteURL(url, hasHeaders, methodUC, requestTimeout);
+  }
+  if (methodUC === 'POST' || methodUC === 'PUT' || methodUC === 'PATCH') {
+    snippetRequest = getSnippetPostPutOrPatchForm(url, hasParams, hasHeaders, methodUC, mode, requestTimeout);
+  }
+  return snippetRequest;
```
Comment (LuisTejedaS) [inline codegens/r-httr/lib/rHttr.js:224]: What about other verbs like options head etc.

File: codegens/r-httr/lib/util/parseBody.js [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    try {
+      let jsonBody = JSON.parse(body);
+      bodySnippet += `'${JSON.stringify(jsonBody, null, indentation.length)}';`;
+    }
+    catch (error) {
+      bodySnippet += `"${sanitizeString(body.toString(), bodyTrim)}"`;
+    }
+  }
+  else {
+    // bodySnippet += `"${sanitizeString(body.toString(), bodyTrim)}"`;
```
Comment (LuisTejedaS) [inline codegens/r-httr/lib/util/parseBody.js:106]: This line is no longer needed we can remove it

File: codegens/r-httr/lib/util/parseBody.js [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+ *  Takes in a key value form data and creates the PHP guzzle structure
+ * for files
+ *
+ * @param {Object} data item from the array of form data (key value).
+ * @param {String} indentation indentation to be added to the snippet
+ * @param {boolean} bodyTrim trim body option
+ * @returns {String} snippet of the body generation
+ */
+function buildFormDataParamFile (data, indentation, bodyTrim) {
+  // let name = `${indentation.repeat(2)}[\n${indentation.repeat(3)}` +
```
Comment (LuisTejedaS) [inline codegens/r-httr/lib/util/parseBody.js:170]: This code is no longer needed

File: codegens/r-httr/npm/test.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+  exit = require('shelljs').exit,
+  prettyms = require('pretty-ms'),
+  startedAt = Date.now(),
+  name = require('../package.json').name;
+
+require('async').series([
+  require('./test-lint'),
+  require('./test-newman')
+  // Add a separate folder for every new suite of tests
+  // require('./test-unit')
```
Comment (LuisTejedaS) [inline codegens/r-httr/npm/test.js:12]: We should be running unit testing when we run npm run test in your folder

File: codegens/r-httr/package.json [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{
+  "name": "@postman/codegen-r-httr",
+  "version": "0.0.1",
+  "description": "Converts Postman SDK Requests to swift code snippet",
```
Comment (LuisTejedaS) [inline codegens/r-httr/package.json:4]: Change swift for r  or r-httr

File: codegens/r-httr/package.json [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    "syntax_mode": "r"
+  },
+  "directories": {
+    "lib": "lib",
+    "test": "test"
+  },
+  "scripts": {
+    "test": "node npm/test.js",
+    "test-lint": "node npm/test-lint.js",
+    "test-newman": "node npm/test-newman.js"
```
Comment (LuisTejedaS) [inline codegens/r-httr/package.json:19]: missing test unit command?

File: codegens/r-httr/test/unit/convert.test.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+var expect = require('chai').expect,
+  { convert } = require('../../index');
+  // sdk = require('postman-collection'),
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:3]: Unnecessary commented code

File: codegens/r-httr/test/unit/convert.test.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  // fs = require('fs'),
+  // path = require('path');
+  // collectionsPath = './fixtures';
+
+describe('convert function', function () {
+
+  // it('should convert a simple get request', function (done) {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:10]: This test is comented

**Review by LuisTejedaS — CHANGES_REQUESTED** [review]
You are missing test for options e.g. how do you know that time out snippet is present

File: codegens/r-httr/test/unit/convert.test.js [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    done();
+  });
+
+  it('should convert a simple get request with timeout', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:26]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[1].request, { requestTimeout: 3 }, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      expect(snippet.includes('timeout(3)')).to.be.true;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:35]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    done();
+  });
+
+  it('should convert a post request with formdata', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:42]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[4].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:50]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+    done();
+  });
+
+  it('should convert a post request with raw data', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:57]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[6].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:65]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
+    convert(collection.items.members[7].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:80]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+    done();
+  });
+
+  it('should convert a post request with urlencoded', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:72]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
+    done();
+  });
+
+  it('should convert a post request with json with raw', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:87]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[8].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:95]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+    done();
+  });
+
+  it('should convert a post request with javascript with raw', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:102]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[9].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:110]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 23]
```diff
@@ -0,0 +1,148 @@
...
+    done();
+  });
+
+  it('should convert a post request with xml with raw', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:117]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[10].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:125]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+    done();
+  });
+
+  it('should convert a post request with binary file', function (done) {
+    const collection = new sdk.Collection(JSON.parse(
+      fs.readFileSync(path.resolve(__dirname, './fixtures/sample_collection.json').toString())));
+    // collection.items.members.forEach((item) => {
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:132]: unnecessary  commented line

File: codegens/r-httr/test/unit/convert.test.js [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+    convert(collection.items.members[25].request, {}, function (err, snippet) {
+      if (err) {
+        console.error(err);
+      }
+      expect(snippet).to.not.be.empty;
+      fs.writeFileSync(path.join(__dirname, './fixtures/snippet.r'), snippet);
+    });
+    // });
```
Comment (LuisTejedaS) [inline codegens/r-httr/test/unit/convert.test.js:140]: unnecessary  commented line

**Review by LuisTejedaS — CHANGES_REQUESTED** [review]
Some code still commented

**Review by LuisTejedaS — APPROVED** [review]
LGTM

**Comment by erfemega** [comment]
Hi @umeshp7 I add some screenshots from the generated code:
Rquest headers with special characters:
image

Request Headers:
image

Request headers with disabled headers:
image

GET Request with disabled query:
image

POST Raw text:
image

POST json with raw:
image

POST GraphQL:
image

Javascript with Raw:
image

Text/Xml with raw:
image

POST Url encoded data:
image

PUT Request:
image

PATCH Request:
image

DELETE Request:
image

Post empty form data:
image

POST Multipart/form-data with text parameters:
image

Single/multiple file load via form-data:
image

Binary file upload via form data:
image

Multiple headers with same value:
image

**Comment by webholik** [comment]
image
@erfemega Can you fix the text alignment in line number 3 in the above image. `'pl' = '\'a\''` should be aligned with the next line `'qu' = '"b"'`.

**Comment by umeshp7** [comment]
Also we need appropriate spacing. For example: 
snippet-r

Here the spacing should be something like this:

```
library(httr)

body = list(
    'pl' = 'a'
)

res <- VERB(...)

cat(content())
```

This way its a bit more readable.

**Comment by erfemega** [comment]
Sure @webholik it was my mistake when I was copying the snippets, the generated snippet does not have this issue.
Done @umeshp7 
I've edited my last comment with the right snippets.
Thank you

**Pull request merged**